### ISSUE TITLE
AUTH_GETCURRENT yields undefined

### ISSUE THREAD

**Issue by RWOverdijk** [benwinding/react-admin-firebase#23]
I have a startup saga that should execute when the user is logged in. I try to check this by using the AUTH_GETCURRENT action, but the result is always undefined.

**lib/providers.ts**

```ts
import { FirebaseAuthProvider, FirebaseDataProvider } from 'react-admin-firebase';
import addUploadFeature from './addUploadFeature';

const config = {
  /* ... */
};

export const dataProvider = addUploadFeature(FirebaseDataProvider(config));
export const authProvider = FirebaseAuthProvider(config);
```

**lib/sagas.ts**

```ts
import { crudGetList, showNotification } from 'react-admin';
import { authProvider } from './providers';

export default function* sagas() {
  const identity = yield authProvider('AUTH_GETCURRENT');
  console.log({ identity }); // { identity: undefined }
}
```

**Comment by RWOverdijk** [comment]
I found out that this is because you don't return the user. Also, you're not using the unsubscribe method provided by the `onAuthStateChanged`. I really need this stuff so I don't mind helping, but there's a lot of basic typescript stuff wrong, too. Perhaps it's easier if we talk on a chat somewhere 😄

**Comment by RWOverdijk** [comment]
I am working around it using this code:

```ts
import firebase from 'firebase/app';
import 'firebase/auth';
import 'firebase/firestore';
import { AUTH_GET_PERMISSIONS } from 'react-admin';

export default (app: firebase.app.App, provider: any) => {
  const db = app.firestore();
  const auth = app.auth();

  const getUser = (): Promise<firebase.User | void> => new Promise(resolve => {
    const unsubscribe = auth.onAuthStateChanged((user) => {
      unsubscribe();

      resolve(user || undefined);
    });
  });

  return async (type: string, params: {}) => {
    if (type === AUTH_GET_PERMISSIONS) {
      const user = await getUser();

      if (!user) {
        return false;
      }

      const docRef: firebase.firestore.DocumentReference = db.collection('users').doc(user.uid);

      try {
        const doc: firebase.firestore.DocumentSnapshot = await docRef.get();
        const data = doc.data();

        if (!doc.exists || !data) {
          return false;
        }

        return data.role;
      } catch (error) {
        return false;
      }
    }

    return provider(type, params);
  };
};
```

Which I use like so:

```ts
export const app = firebaseApp.apps.length ? firebaseApp.app() : firebaseApp.initializeApp(config);
export const authProvider = addPermissions(app, FirebaseAuthProvider(config));
```

It works by having a collection called "user" where the id is the user's uid. Don't forget to add rules to your firestore setup:

```
service cloud.firestore {
  match /databases/{database}/documents {
    match /users/{userId} {
			allow read: if belongsTo(userId);
      allow write: if false;
    }
  }
}

function belongsTo(userId) {
    return request.auth.uid == userId
}
```

The user document now has one key called "role" in firebase, with the string value "admin", allowing me to do this in for example App.tsx:

```tsx
class App extends React.Component {
  public render() {
    return (
      <Admin
        customSagas={[dataSaga]}
        dataProvider={dataProvider}
        authProvider={authProvider}
      >
        {(permissions: string) => [
          <Resource name='items' icon={ArtTrackIcon} list={ItemList} show={ItemShow} create={ItemCreate} edit={ItemEdit}/>,
          <Resource name='locales' icon={LanguageIcon} list={LocaleList} show={LocaleShow} create={LocaleCreate} edit={LocaleEdit}/>,
          permissions === 'admin' ? <Resource name='config' list={ConfigList} create={ConfigCreate} edit={ConfigEdit} /> : null,
        ]}
      </Admin>
    );
  }
}
```

**Comment by benwinding** [comment]
Hey @RWOverdijk,

Apologies for the delay, I've been trying to get this package in a better state to include unit-tests for more reliable releases.

### Problem

This has been confusing me. Looks like I still have **a lot to learn** about:

- Redux sagas
- JavaScript generator functions: `function*` and `yield`
- The `react-admin` `authProvider` and how it wraps around the resources or something.
- ... _many other things_

> I found out that this is because you don't return the user.

From what I can gather, you need the `FirebaseAuthProvider` to return the result of `AUTH_GETCURRENT` action... (_as it's meant too :disappointed:_).

### Solution

After messing around with the `FirebaseAuthProvider` I found that I was only awaiting the promises, not returning them. So I fixed that (commit: fd6a0fc5d58903eab9c3edfb5c07cbf9825beb32.) and here's a minimal working example:

``` javascript
const authProvider = FirebaseAuthProvider(firebaseConfig, options);

function* testSaga() {
  const identity = yield authProvider('AUTH_GETCURRENT');
  console.log({ identity }); // { identity: undefined }
}
```
It seems you've solved this solution already using your `addPermissions` wrapper, but if you need this it's implemented in version `0.6.5` 
Thanks again,
Ben
